Create separate metadata class

sjvenditto opened the pull request:
## Summary
New private class: `_MetadataMixin` (`core/metadata_class.py`). Can be inherited by:
- `IntervalSet`
- `TsdFrame`
- `TsGroup`
This class assumes that whatever is inheriting it has the private property `self._initialized`.
### Properties
- `_metadata`: pandas DataFrame containing metadata. The row index corresponds to:
	- `IntervalSet`: interval index (`self.index`)
	- `TsdFrame`: columns names (`self.columns`)
	- `TsGroup`: group key (`self.index`)
- **`metadata`: public read-only view of metadata**
- `metadata_columns`: list of metadata columns (same as `list(_metadata.columns))
- `metadata_index`: index of metadata rows (same as `_metadata.index`)

### Methods
- `__dir__`: adds metadata columns to `dir` for tab completion
- `__setattr__`: allows metadata columns to be set as attribute after initialization (`self._initialized = True`)
- `__getattr__`: allows for accessing metadata as an attribute, e.g. for `obj` with metadata `label`: `obj.label`
- `__setitem__`: allows for setting metadata via key. e.g. `obj["label"] = values`. calls `set_info`
- `__getitem__`: allows for getting metadata via key, e.g. `obj["label"]`. calls `get_info`
- `_raise_invalid_metadata_column_name(name)`: handles error catching for invalid names. invalid names are:
	- non-strings
	- existing attribute names (excluding metadata attributes)
	- existing column names (anything in `self.columns`, if it exists)
	- names that start with a number
	- names with special characters (excluding underscores)
- `_check_metadata_column_names(*args,*kwargs)`: wrapper function for calling `_raise_invalid_metadata_column_name` on each metadata input argument
- `set_info(*args,**kwargs)`: handles parsing input type, name error checking, and adding metadata. pulled from `ts_group.py`
- `get_info(key)`: function for grabbing a metadata column via input key from dataframe. also pulled from `ts_group.py`

## General features
- **metadata can be set at object initialization by passing a dictionary or dataframe with the `metadata` input argument. `TsGroup` metadata can also be initialized at initiation with variable kwargs for backwards compatibility.**
- metadata can be set with `set_info`, via key (`obj["label"]=values`), or via attribute (`obj.label = values`)
- metadata can be accessed by `get_info` via key (`obj["label"]`) or via attribute (`obj.label`)
- multiple metadata columns can be accessed at once via a list of keys (`obj[["label1","label2"]]`)
- stored metadata can be any type

## Updates to all classes
- `save` and `_from_npz_reader` accounts for metadata
- updates to `__repr__`
- updates to `__setitem__` to bypass immutability (if applicable) and allow for setting metadata via string keys
- updates to `__getitem__` to allow getting metadata via string keys
- adding `__setattr__` to allow setting metadata attributes

## Class-specific changes
Excludes general updates mentioned above

### IntervalSet
**`_jitted_functions.py`**
- `jitintersect`: track metadata index
- `jitdiff`: track metadata index

 **`interval_set.py`** 
- added private property `_initialized` for metadata attribute setter
- when constructing with `pd.DataFrame`, add additional columns as metadata
- `__getitem__` also allows for indexing multiple metadata columns for specific intervals, e.g. `iset[0,["label1","label2"]` or `iset[[1,2],["label1","label2"]`, etc.
- `intersect` preserves metadata
- `set_diff` preserves metadata
- `as_dataframe` appends metadata
- `split` splits metadata
- `union`, `merge_close_intervals`, and `time_span` issue warnings that metadata is dropped

**`utils.py`**
- update to `_IntervalSetSliceHelper > __getitem__` to also check for metadata column names

**`interface_nwb.py`**
- `_make_interval_set` updated to be able to load in metadata from nwb files

### TsdFrame
**`base_class.py`**
- pass metadata to kwargs whenever "columns" are passed to kwargs (`value_from` and `restrict`)
 
 **`time_series.py`**
- pass metadata to kwargs whenever "columns" are passed to kwargs (`__array_ufunc__`, `__array_function__`, `bin_average`,`dropna`,)
- special case in `__getitem__` to make `obj['label'] == obj[:,'label']` for consistency in accessing metadata
- `copy` updated to preserve metadata and column names

## Tests
Most new tests added in `test_metadata.py`

### Class-agnostic tests
Parameterized tests to work on each of `IntervalSet`, `TsdFrame`, and `TsGroup` objects
- `test_add_metadata`: test to add metadata with `set_info` using keyword arguments
- `test_add_metadata_key`: test to add metadata via key assignment
- `test_add_metadata_attr`: test to add metadata via attribute assignment
- **`test_add_metadata_many`: test multiple kwargs inputs into `set_info`**
- `test_add_metadata_df`: test to add metadata with `set_info` using a  `pd.DataFrame` as a positional argument
- `test_add_metadata_error`: error testing for invalid metadata names & invalid shape
- `test_add_metadata_key_error`: error testing on adding metadata with keys with class-specific error messages.
- `test_overwrite_metadata`: test that metadata columns are mutable with all setters
- `test_metadata_slicing`: tests the ability to slice an object with specific metadata value. e.g. `obj[obj.label == 1]` (this also includes some type-specific checks)
- `test_metadata_index_columns`: checks for single and multiple metadata column indexing
- `test_save_and_load_npz`: checks when an object is saved as an npz, metadata is saved as an additional key `"_metadata"`. checks loading of saved metadata
- `test_no_conflict_between_intervalset_and_metadatabase`, `test_no_conflict_between_tsdframe_and_metadatabase`, and `test_no_conflict_between_tsgroup_and_metadatabase` from Edoardo to test for double inheritance

### Class-specific tests
#### IntervalSet
Most metadata tests currently contained in `test_metadata.py`. These are modeled after existing tests for interval sets without metadata
- `test_create_iset_with_metadata`
- `test_create_iset_from_df_with_metadata`
- `test_get_iset_with_metadata`: test slices that should return an intervalset
- `test_slice_iset_with_metadata`: tests various vertical and horizontal slicing methods
- `test_get_iset_with_metadata_errors`: check errors for invalid indexing
- `test_as_dataframe_metadata`
- `test_intersect_metadata`
- `test_set_diff_metadata`
- `test_drop_short_intervals_metadata`
- `test_drop_long_intervals_metadata`
- `test_split_metadata`
- `test_drop_metadata_warnings`

Tests changed in `test_interval_set.py`:
- **`test_save_npz` to test saving with and without metadata**

#### TsdFrame
Tests in `test_metadata.py`
- `test_tsdframe_metadata_slicing`: parameterized test to ensure that `obj[obj.label==a]` is the same as `obj[:,obj.label==a]`, specific for `TsdFrame`
- `test_tsdframe_add_metadata_error`: specific test for error adding metadata with names that match tsdframe columns

These are changes to tests in `test_time_series.py`
- `test_create_tsdframe` has been parameterized to include cases with metadata
- `class Test_Time_Series_3` has 3 additional cases added: 1) with str column names, 2) metadata without str column names 3) metadata with str column names
	- new test for `TsdFrame.copy()`
	- `test_horizontal_slicing` parameterized with checks for metadata
	- `test_vertical_slicing` parameterized with checks for metadata
	- `test_str_indexing` parameterized with checks for metadata. is redundant when column names are not strings.
	- `test_save_npz` updated to check for saved metadata. possibly redundant with class-agnostic test
	- **`test_vert_and_horz_slicing` for more variations of slicing tests**

#### TsGroup
No new metadata specific tests added or moved from `test_ts_group.py`. Modified tests:
- `test_save_npz` updated to work with new way of saving metadata
- `test_get_rate` no longer tests for "freq" and "frequency"
- ~~`test_add_metainfo_raise_error` and `test_keys` updated to reflect new error strings~~
- **redundant metadata tests removed**

### Other modified tests
- `test_jitted.py` > `test_jitintersect` and `test_jitdiff` both additionally test metadata
- `test_npz_file.py > test_load_tsdframe` also checks for metadata (possibly redundant)
- `test_nwb.py > test_add_TimeIntervals`: replace check for warning with check of loaded metadata

## Issues
- Since it was copied from `ts_group.py`, documentation for `set_info` only has examples specific to `TsGroup`
	- examples for adding metadata is lacking in general
- ~~`get_info` has a key name check specific for `TsGroup`~~
- ~~`TsdFrame` is mutable via integer keys~~
- ~~`IntervalSet` has more flexible indexing of metadata than other objects~~
- ~~no warnings when metadata is dropped for `IntervalSet`~~
	- ~~`union`~~
	- ~~`merge_close_intervals`~~
- ~~additional tests for metadata in `IntervalSet`~~
	- ~~`drop_short_intervals` and `drop_long_intervals`: test metadata is preserved~~
	- ~~`split`: test metadata is split~~
- ~~adding checks for double inheritance~~
- ~~no tests for updates to `jitintersect` and `jitdiff`~~

BalzaniEdoardo commented on pynapple/core/time_series.py:
```diff
@@ -974,6 +990,13 @@ def round_if_float(x):
                 return tabulate([], headers=headers) + "\n" + bottom
 
+    def __getattr__(self, name):
+        # wrap in try except to avoid infinite recursion from metadata pickling
+        try:
+            return MetadataBase.__getattr__(self, name)
+        except:
+            return super().__getattr__(name)
```
this is a solution that avoids running the recursion
```python
        # avoid infinite recursion when pickling due to
        # self._metadata.column having attributes '__reduce__', '__reduce_ex__'
        if name in ("__getstate__", "__setstate__", "__reduce__", "__reduce_ex__"):
            raise AttributeError(name)
```

Another cleaner solution, was to reimplement `__getstate__` and `__setstate__`, as in here 

https://github.com/flatironinstitute/nemos/blob/main/src/nemos/basis.py

sjvenditto replied:
I added in this solution!

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -22,12 +22,12 @@ def __init__(self, *args, **kwargs):
-            self._metadata_index = self.columns
+            self.metadata_index = self.columns
         else:
             # what if index is not defined?
```
not in any of the objects currently supported

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,236 @@
+            for arg in args:
+                if isinstance(arg, pd.DataFrame):
+                    if pd.Index.equals(self._metadata.index, arg.index):
+                        object.__setattr__(self, "_metadata", self._metadata.join(arg))
```
this doesn't work for `TsdFrame` because `setattr` is "turned off" after initialization to enforce immutability

BalzaniEdoardo replied:
Makes sense!

BalzaniEdoardo replied:
Unrelated to this PR, but I was thinking to be 100% immutable, we should also raise an error in `__delattr__`; I am not sure why one would do that, but you can still remove attrs right now; @gviejo If you want I can add that functionality

```python
# this should raise an error but currently doesn't
a = nap.Tsd(np.arange(10), np.arange(10))
delattr(a, "index")
```

sjvenditto replied:
Another solution is to update the columns instead of replacing `_metadata`
```
self._metadata[arg.columns] = arg
```

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -22,12 +22,12 @@ def __init__(self, *args, **kwargs):
         """
         if self.__class__.__name__ == "TsdFrame":
```
`IntervalSet` also has a "columns" attribute

BalzaniEdoardo replied:
I had no clue! Is it new? Or since forever?

sjvenditto replied:
no idea! but it caused an issue when I tried your solution

gviejo commented on pynapple/core/interval_set.py:
```diff
@@ -533,7 +550,7 @@ def drop_short_intervals(self, threshold, time_units="s"):
             np.array([threshold], dtype=np.float64), time_units
         )[0]
-        return self[(self.values[:, 1] - self.values[:, 0]) > threshold]
+        return self[(self.values[:, 1] - self.values[:, 0]) > threshold, :]
```
Why is it needed?

sjvenditto replied:
this might be an artifact to before I changed the indexing, where the second slice was to grab the metadata columns. I don't think it's necessary now

sjvenditto replied:
I took out this slice (and the slice in drop_long_intervals)

gviejo commented on pynapple/core/ts_group.py:
```diff
@@ -187,66 +189,21 @@ def __init__(
-                    raise
-            else:
-                self.set_info(**{key: value})
+            # do we want metadata set outside of `set_info` to be allowed?
```
Yes it should be possible

sjvenditto commented on pynapple/core/interval_set.py:
```diff
@@ -255,32 +226,70 @@ def __getitem__(self, key, *args, **kwargs):
-                elif key[1] == slice(None, None, None) or key[1] == slice(0, 2, None):
-                    output = self.values.__getitem__(key)
-                    return IntervalSet(start=output[:, 0], end=output[:, 1])
+                df = self.as_dataframe()
```
this is to be flexible with indexing the metadata columns. we could separate out cases of indexing metadata columns and not, to make indexing just the start and end columns not rely on it

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,242 @@
+import pandas as pd
+
+
+class MetadataBase:
```
I changed to the second suggestion for now, unless you think changing to an abstract class is better

gviejo replied:
I am ok with this. Maybe @BalzaniEdoardo has a different opinion about this one?

BalzaniEdoardo replied:
I would define an abstract class only if we think we need to subclass it and the subclass must reimplement some methods.

For example, if we have a Tsd vs TsGroup metadata method that must differ in the implementation. Otherwise private is fine.

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,242 @@
+        if key in self._metadata.columns:
+            return self.get_info(key)
+        else:
+            raise KeyError(r"Key {} not in group index.".format(key))
```
I moved this error back to TsGroup (where it was pulled from)

gviejo commented:
Here is a summary of the behavior we would like to have across objects. Let me know if you agree @sjvenditto @BalzaniEdoardo 

#### Supported objects
obj in {IntervalSet, TsGroup, TsdFrame}

#### Passing metadata
`obj(args, a=[1,2,3], b=['c','d','f'])`  # during initialization
`obj.set_info(c=[3,2,1])`                   # with set_info
`obj['d'] = ['i', 'j', 'k'] `                        # as a key of a dict

#### Getting metadata
`obj['a']`            -> [1,2,3]          # as a key of a dict
`obj.a`               -> [1,2,3]          # as an attribute
`obj.get_info('a')`   -> [1,2,3]          # with get_info

#### Slicing with metadata
`obj[obj.a == 1]`

#### Rules
- Label names should always be string
- For TsdFrame, metadata columns should not conflict with tsdframe columns.

#### Issues:
- obj[obj.a == 1] should work as obj[:,obj.a == 1]. Do we want to support the first syntax?
- Should `getby_category`, `getby_intervals` and `getby_threshold` be part of Metadata class?

sjvenditto commented:
This looks like a good list to me. I think the only things missing right now is using keys to get and set metadata in `TsdFrame` as well as slicing with particular metadata values. 

One idea was to separate this pull request as getting the metadata object in place, and then using another pull request to add in the grouping functionality. Both issues you pointed out (as well as slicing with particular metadata values) might make sense for the next pull request?

gviejo commented:
Yes i agree for doing the grouping somehwere else.

BalzaniEdoardo commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,242 @@
+                print(k)
+                invalid_cols += [k]
+
+        if invalid_cols:
```
It seems that it checks if a key in the kwargs dict is already in the col names and of it is, then it adds the key to the invalid. I forgot why thee is q check associated to the values too

sjvenditto commented on pynapple/core/interval_set.py:
```diff
@@ -118,16 +121,16 @@ def __init__(self, start, end=None, time_units="s"):
-                    - column names are ["start", "end"]                    
+                DataFrame must contain columns name "start" and "end" for start and end times.                   
                 """
+            m = start.drop(columns=["start", "end"])
```
fixed this

sjvenditto commented:
I updated the slicing now, so passing, getting and slicing with metadata should be consistent across the three objects. For the first issue, I did support the first syntax by checking for the condition:
```
if isinstance(key, pd.Series) and key.index.equals(self.columns):
      # if indexing with a pd.Series from metadata, transform it to tuple with slice(None) in first position
      key = (slice(None, None, None), key)
```                

`IntervalSet` still has some unique behavior from stuff I implemented before. E.g. you can get multiple columns at a time, like `ep[:,['meta1','meta2']]` would return a dataframe with `'meta1'` and `'meta2'` columns. However `ep[['meta1','meta2']]` is not currently supported. Is this something that would be worthwhile to extend to the other objects, as well as including a case like `ep[['meta1','meta2']]`?

BalzaniEdoardo commented on pynapple/core/time_series.py:
```diff
@@ -915,70 +930,119 @@ def __init__(
     def loc(self):
         return _TsdFrameSliceHelper(self)
 
     def __repr__(self):
```
this method for repr it is quite annoying, maybe it is worth moving it into a script, which contains the funcitons for computing the  reprs of all the objects. maybe there is some shared code that help compressing

BalzaniEdoardo reviewed (commented):
left some comments. still did not go through the tests

BalzaniEdoardo commented on pynapple/core/interval_set.py:
```diff
@@ -75,12 +76,12 @@
 
 
-class IntervalSet(NDArrayOperatorsMixin):
+class IntervalSet(NDArrayOperatorsMixin, _MetadataBase):
```
except for start and end, we should be able to add metadata

sjvenditto replied:
whoops you are correct, I'll add in this functionality

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,250 @@
+                    if pd.Index.equals(self._metadata.index, v.index):
+                        self._metadata[k] = v
+                    else:
+                        print(self._metadata.index)
+                        print(v.index)
+                        print(v)
```
this is an artifact of the `set_info` method I copied from `ts_group.py`. I can remove it

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,250 @@
+            raise ValueError(
+                f"Invalid metadata name '{name}'. Metadata name cannot start with a number"
+            )
+        if (" " in name) or ("." in name):
```
I can change it to disallow any special characters

sjvenditto replied:
Except for underscores, which can be allowed

gviejo commented on tests/test_time_series.py:
```diff
@@ -741,143 +790,215 @@ def test_to_tsgroup(self, tsd):
+        np.testing.assert_array_almost_equal(file["end"], tsd.time_support.end)
 
         # Path("tsd.npz").unlink()
         # Path("tsd2.npz").unlink()
```
```suggestion
        Path("tsd2.npz").unlink()
```

gviejo reviewed (changes requested):
In interface_nwb.py line 105 to 120 to me it should be removed.

sjvenditto commented on pynapple/io/interface_nwb.py:
```diff
@@ -119,12 +119,12 @@ def _make_interval_set(obj, **kwargs):
-                    stacklevel=2,
-                )
-                return df  # Too many metadata to split the epoch
+            else:  # assume extra columns are metadata
```
removed

sjvenditto commented on tests/test_npz_file.py:
```diff
@@ -30,96 +30,112 @@
-    }
+        }
+    ),
+    "iset": nap.IntervalSet(start=np.array([0.0, 5.0]), end=np.array([1.0, 6.0])),
```
I do have some testing for this in `test_metadata.py`, but since I had added the metadata test for tsdframe here, I updated it to test all three and to test versions with and without metadata (tsgroup always has metadata, so it tests with and without an extra field). I changed the tests such that metadata only gets checked in `test_load` to reduce some redundancy

BalzaniEdoardo commented on tests/test_metadata.py:
```diff
@@ -0,0 +1,705 @@
+"""Tests for metadata in IntervalSet, TsdFrame, and TsGroup"""
+
+from numbers import Number
```
General comment for these tests, we should have at least more than 1 case for each operation, trying to catch corner cases like:

threshold for splitting intervals sts larger than any given epoch, smaller than each epoch, inbetween... slice 0 element, 1 element, slice with step =1 step >1... stuff like that. For the rest they look good

BalzaniEdoardo replied:
Basically,my main point is try potential corner cases: add metadata of len=1, len>1, len=0, pass a list, an array etc of all these sizes

BalzaniEdoardo reviewed (changes requested):
Sarah, this is shaping in a very powerful module, so thanks for this! I left some comments, but the major ones are:

- change how metadata is passed at init
- change what can be used as metadata col names, and be as permissive as possible
- simplify how the attributes are retrieved, try first non-metadata, if fails, get metadata

sjvenditto commented on pynapple/core/interval_set.py:
```diff
@@ -186,101 +191,202 @@ def __init__(self, start, end=None, time_units="s"):
+            # self[array_like]
             output = self.values.__getitem__(key)
-            return IntervalSet(start=output[:, 0], end=output[:, 1])
+            metadata = _MetadataBase.__getitem__(self, key).reset_index(drop=True)
```
interval set index is always reset to 0, so if it isn't reset, it won't match the interval set index

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,278 @@
+                    if len(self._metadata.index) == len(v):
+                        self._metadata[k] = np.asarray(v)
+                    else:
+                        raise RuntimeError(
```
updated to ValueError

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,278 @@
+                    if pd.Index.equals(self._metadata.index, arg.index):
+                        self._metadata[arg.columns] = arg
+                    else:
+                        raise RuntimeError("Metadata index does not match")
```
updated to ValueError

sjvenditto commented on pynapple/core/interval_set.py:
```diff
@@ -186,101 +191,202 @@ def __init__(self, start, end=None, time_units="s"):
+            # easiest to convert to dataframe and then slice
+            # in case of mixing ["start", "end"] with metadata columns
+            df = self.as_dataframe()
+            if all(x in key[1] for x in ["start", "end"]):
```
this is a mistake, I changed it to just `key`

sjvenditto commented on tests/test_time_series.py:
```diff
@@ -1477,16 +1689,72 @@ def test_pickling(obj):
+        (
+            1,
+            3,
+            "banana",
```
very important!

sjvenditto commented on tests/test_time_series.py:
```diff
@@ -741,143 +790,215 @@ def test_to_tsgroup(self, tsd):
+            assert np.all(tsdframe[index].metadata_columns == tsdframe.metadata_columns)
+            assert np.all(tsdframe[index].metadata_index == tsdframe.metadata_index)
+
+    @pytest.mark.parametrize("index", [0, [0, 2]])
```
I've added in additional slicing & indexing tests

sjvenditto commented on tests/test_time_series.py:
```diff
@@ -741,143 +790,215 @@ def test_to_tsgroup(self, tsd):
+                tsdframe[:, index].metadata_index == tsdframe.metadata_index[index]
+            )
+
+    @pytest.mark.parametrize("index", [slice(0, 10), [0, 2]])
```
I've added in additional slicing & indexing tests

sjvenditto commented on tests/test_metadata.py:
```diff
@@ -0,0 +1,705 @@
+        ],
+    )
+    class Test_Add_Metadata:
+        def test_add_metadata(self, obj, info):
```
After adding the kwargs initialization, I added a test for adding multiple metadata with `set_info`

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,278 @@
+                f"'{type(self).__name__}' object has no attribute '{name}'"
+            )
+
+    def __setitem__(self, key, value):
```
I've updated this behavior

sjvenditto commented on tests/test_interval_set.py:
```diff
@@ -448,108 +504,118 @@ def test_jitfix_iset_random():
     start = np.around(np.array([0, 10, 16], dtype=np.float64), 9)
     end = np.around(np.array([5, 15, 20], dtype=np.float64), 9)
-    ep = nap.IntervalSet(start=start,end=end)
+    ep = nap.IntervalSet(start=start, end=end)
```
added

sjvenditto commented on tests/test_interval_set.py:
```diff
@@ -448,108 +504,118 @@ def test_jitfix_iset_random():
     end = np.around(np.array([5, 15, 20], dtype=np.float64), 9)
     with pytest.warns(UserWarning, match=r"Some epochs have no duration"):
-        nap.IntervalSet(start=start,end=end)
+        nap.IntervalSet(start=start, end=end)
```
added in warning for when metadata gets dropped due to changed intervals at creation

sjvenditto commented on tests/test_interval_set.py:
```diff
@@ -448,108 +504,118 @@ def test_jitfix_iset_random():
+
 
 def test_sort_starts():
     start = np.around(np.array([10, 0, 16], dtype=np.float64), 9)
```
added in warning for when metadata gets dropped due to changed intervals at creation

sjvenditto commented on pynapple/core/interval_set.py:
```diff
@@ -75,12 +76,12 @@
     """
 
-    def __init__(self, start, end=None, time_units="s"):
+    def __init__(self, start, end=None, time_units="s", **kwargs):
```
removed kwargs initialization of metadata for interval set and tsdframe. however, I've left it in place with a FutureWarning for TsGroup for backwards compatibility

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,278 @@
+            # existing non-metadata attribute
+            raise ValueError(
+                f"Invalid metadata name '{name}'. Metadata name must differ from "
+                f"{type(self).__dict__.keys()} attribute names!"
```
this will be handled in the next PR!

sjvenditto commented on pynapple/core/metadata_class.py:
```diff
@@ -0,0 +1,278 @@
+                    if pd.Index.equals(self._metadata.index, v.index):
+                        self._metadata[k] = v
+                    else:
+                        raise RuntimeError(
```
updated to ValueError

BalzaniEdoardo reviewed (approved):
Congrats Sara! Great PR 

🦭 of approval
